Kernel/Time/HardwareTimer.h

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright (c) 2020, Liav A. <liavalb@hotmail.co.il>
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include <AK/Function.h>
+#include <AK/RefCounted.h>
+#include <AK/String.h>
+#include <Kernel/Interrupts/IRQHandler.h>
+
+namespace Kernel {
+
+enum class HardwareTimerType {
+    i8253 = 0x1,                  /* PIT */
+    RTC = 0x2,                    /* Real Time Clock */
+    HighPrecisionEventTimer = 0x3 /* also known as IA-PC HPET */
+};
+
+class HardwareTimer : public RefCounted<HardwareTimer>
+    , public IRQHandler {
+public:
+    virtual HardwareTimerType timer_type() const = 0;
+    virtual const char* model() const = 0;
+    virtual const char* purpose() const override;
+
+    virtual void change_function(Function<void(const RegisterState&)>);
+    virtual size_t ticks_per_second() const = 0;
+
+    virtual bool is_periodic() const = 0;
+    virtual void reset_to_default_ticks_per_second() = 0;
+    virtual bool set_frequency(size_t frequency) = 0;
```

Comment:
Maybe we should use a slightly different name here? set_* implies to me that it returns void, and that the caller needs to validate beforehand. This gave me a thought though, it would be cool if [[nodiscard]] attributes were able to be inherited ☹️

Reply:
maybe `try_to_set_frequency` will be a good name, what do you think? :)

Reply:
Yeah I guess :) I'm not sure aye. I just expect that if this represents a generic hardware timer this could return false a whole lot. Damn hardware only being able to change clock frequencies in steps because of those pesky clock multipliers / dividers :P